From cf11cde6811f3095e28f2d1d4b0f58dfcf468c68 Mon Sep 17 00:00:00 2001
From: Shawn Ferry <sferry@gmail.com>
Date: Fri, 21 Apr 2017 14:18:28 -0400
Subject: [PATCH 1/3] (MODULES-4749) puppetlabs-mysql : Solaris support

(MODULES-5277) puppetlabs-mysql : default version fact is not set on Solaris

This patch adds Solaris support as currently shipped by Oracle. Mysql
version is extracted from `pkg mediator mysql` and used to set path
values.

This patch also adds the structured fact mysql_solaris which holds
values for consumption in params.pp

Patch is submitted upstream
---
 lib/facter/mysql_solaris.rb                        | 20 +++++++
 manifests/params.pp                                | 19 ++++---
 metadata.json                                      |  7 +++
 spec/classes/mycnf_template_spec.rb                |  6 +-
 spec/classes/mysql_bindings_spec.rb                |  6 +-
 spec/classes/mysql_client_spec.rb                  |  6 +-
 spec/classes/mysql_server_account_security_spec.rb |  1 -
 spec/classes/mysql_server_backup_spec.rb           |  6 +-
 spec/classes/mysql_server_monitor_spec.rb          |  6 +-
 spec/classes/mysql_server_mysqltuner_spec.rb       |  1 -
 spec/classes/mysql_server_spec.rb                  |  6 +-
 spec/spec_helper_local.rb                          |  8 +++
 spec/unit/facter/mysql_server_id_spec.rb           |  4 +-
 spec/unit/facter/mysql_solaris_spec.rb             | 65 ++++++++++++++++++++++
 spec/unit/facter/mysqld_version_spec.rb            | 22 +++-----
 16 files changed, 127 insertions(+), 58 deletions(-)
 create mode 100644 lib/facter/mysql_solaris.rb
 create mode 100644 spec/unit/facter/mysql_solaris_spec.rb

diff --git a/lib/facter/mysql_solaris.rb b/lib/facter/mysql_solaris.rb
new file mode 100644
index 0000000..c8f2e02
--- /dev/null
+++ b/lib/facter/mysql_solaris.rb
@@ -0,0 +1,20 @@
+Facter.add('mysql_solaris') do
+  begin
+    # Get the mediated version of mysql and update PATH
+    # Fallback to the previously hard coded version (5.5) if there is no mediator
+    ver = Facter::Util::Resolution.exec('pkg mediator -H mysql').split[2]
+
+    # Above exec doesn't raise an error if the command returns non-zero results
+    ver = ver.nil? ? '5.5' : ver
+  rescue
+    # Just in case this raises an error in the future
+    ver = '5.5'
+  end
+  setcode do
+    {
+      'basedir'          => "/usr/mysql/#{ver}",
+      'major_dot_minor'  => ver,
+      'major_minor'      => ver.tr('.', '')
+    }
+  end
+end
diff --git a/manifests/params.pp b/manifests/params.pp
index cb1e6b2..78af6bd 100644
--- a/manifests/params.pp
+++ b/manifests/params.pp
@@ -32,7 +32,6 @@ class mysql::params {
   $daemon_dev_package_provider = undef
   $xtrabackup_package_name     = 'percona-xtrabackup'
 
-
   case $::osfamily {
     'RedHat': {
       case $::operatingsystem {
@@ -321,15 +320,17 @@ class mysql::params {
     }
 
     'Solaris': {
-      $client_package_name = 'database/mysql-55/client'
-      $server_package_name = 'database/mysql-55'
-      $basedir             = undef
-      $config_file         = '/etc/mysql/5.5/my.cnf'
-      $datadir             = '/var/mysql/5.5/data'
-      $log_error           = "/var/mysql/5.5/data/${::hostname}.err"
-      $pidfile             = "/var/mysql/5.5/data/${::hostname}.pid"
+      $server_service_name = "application/database/mysql:version_${::mysql_solaris[major_minor]}"
+      $config_file         = "/etc/mysql/${::mysql_solaris[major_dot_minor]}/my.cnf"
+      $datadir             = "/var/mysql/${::mysql_solaris[major_dot_minor]}/data"
+      $log_error           = "/var/mysql/${::mysql_solaris[major_dot_minor]}/data/${::hostname}.err"
+      $pidfile             = "/var/mysql/${::mysql_solaris[major_dot_minor]}/data/${::hostname}.pid"
+      $client_package_name = "database/mysql-${::mysql_solaris[major_minor]}/client"
+      $server_package_name = "database/mysql-${::mysql_solaris[major_minor]}"
+      $basedir             = $::mysql_solaris['base_dir']
+      $includedir          = undef
       $root_group          = 'bin'
-      $server_service_name = 'application/database/mysql:version_55'
+      $mysql_group         = 'mysql'
       $socket              = '/tmp/mysql.sock'
       $ssl_ca              = undef
       $ssl_cert            = undef
diff --git a/metadata.json b/metadata.json
index 8b27054..a188277 100644
--- a/metadata.json
+++ b/metadata.json
@@ -51,6 +51,13 @@
         "12"
       ]
     },
+    {
+      "operatingsystem": "Solaris",
+      "operatingsystemrelease": [
+        "11",
+        "12"
+      ]
+    },
     {
       "operatingsystem": "Debian",
       "operatingsystemrelease": [
--- puppetlabs-mysql-5.1.0/spec/classes/mycnf_template_spec.rb.orig     2018-01-22 01:03:50.568265110 +0000
+++ puppetlabs-mysql-5.1.0/spec/classes/mycnf_template_spec.rb  2018-01-22 01:04:25.818196954 +0000
@@ -4,9 +4,7 @@
   context 'my.cnf template' do
     on_supported_os.each do |os, facts|
       context "on #{os}" do
-        let(:facts) do
-          facts.merge(root_home: '/root')
-        end
+        let(:facts) { facts }

         context 'normal entry' do
           let(:params) { { override_options: { 'mysqld' => { 'socket' => '/var/lib/mysql/mysql.sock' } } } }
--- puppetlabs-mysql-5.1.0/spec/classes/mysql_bindings_spec.rb.orig     2018-01-22 01:07:44.370207280 +0000
+++ puppetlabs-mysql-5.1.0/spec/classes/mysql_bindings_spec.rb  2018-01-22 01:08:05.737121331 +0000
@@ -3,9 +3,7 @@
 describe 'mysql::bindings' do
   on_supported_os.each do |os, facts|
     context "on #{os}" do
-      let(:facts) do
-        facts.merge(root_home: '/root')
-      end
+      let(:facts) { facts }

       let(:params) do
         {
--- puppetlabs-mysql-5.1.0/spec/classes/mysql_client_spec.rb.orig       2018-01-22 01:35:34.254065666 +0000
+++ puppetlabs-mysql-5.1.0/spec/classes/mysql_client_spec.rb    2018-01-22 01:36:17.313665055 +0000
@@ -3,9 +3,7 @@
 describe 'mysql::client' do
   on_supported_os.each do |os, facts|
     context "on #{os}" do
-      let(:facts) do
-        facts.merge(root_home: '/root')
-      end
+      let(:facts) { facts }

       context 'with defaults' do
         it { is_expected.not_to contain_class('mysql::bindings') }
--- puppetlabs-mysql-5.1.0/spec/classes/mysql_server_backup_spec.rb.orig        2018-01-22 01:10:58.372740539 +0000
+++ puppetlabs-mysql-5.1.0/spec/classes/mysql_server_backup_spec.rb     2018-01-22 01:11:26.630895042 +0000
@@ -8,9 +8,7 @@
           class { 'mysql::server': }
         EOF
       end
-      let(:facts) do
-        facts.merge(root_home: '/root')
-      end
+      let(:facts) { facts }

       let(:default_params) do
         { 'backupuser'         => 'testuser',
--- puppetlabs-mysql-5.1.0/spec/classes/mysql_server_monitor_spec.rb.orig       2018-01-22 01:12:24.097935348 +0000
+++ puppetlabs-mysql-5.1.0/spec/classes/mysql_server_monitor_spec.rb    2018-01-22 01:12:45.508721267 +0000
@@ -2,9 +2,7 @@
 describe 'mysql::server::monitor' do
   on_supported_os.each do |os, facts|
     context "on #{os}" do
-      let(:facts) do
-        facts.merge(root_home: '/root')
-      end
+      let(:facts) { facts }

       let :pre_condition do
         "include 'mysql::server'"
--- puppetlabs-mysql-5.1.0/spec/classes/mysql_server_mysqltuner_spec.rb.orig    2018-01-22 01:13:36.087969552 +0000
+++ puppetlabs-mysql-5.1.0/spec/classes/mysql_server_mysqltuner_spec.rb 2018-01-22 01:14:03.889888284 +0000
@@ -4,8 +4,7 @@
   on_supported_os.each do |os, facts|
     context "on #{os}" do
       let(:facts) do
-        facts.merge(staging_http_get: 'curl',
-                    root_home: '/root')
+        facts.merge(staging_http_get: 'curl')
       end

       context 'ensure => present' do
--- puppetlabs-mysql-5.1.0/spec/classes/mysql_server_spec.rb.orig       2018-01-22 01:38:05.386630188 +0000
+++ puppetlabs-mysql-5.1.0/spec/classes/mysql_server_spec.rb    2018-01-22 01:38:30.251640737 +0000
@@ -3,9 +3,7 @@
 describe 'mysql::server' do
   on_supported_os.each do |os, facts|
     context "on #{os}" do
-      let(:facts) do
-        facts.merge(root_home: '/root')
-      end
+      let(:facts) { facts }

       context 'with defaults' do
         it { is_expected.to contain_class('mysql::server::install') }
--- puppetlabs-mysql-5.1.0/spec/spec_helper_local.rb.orig       2018-01-22 01:15:20.077586849 +0000
+++ puppetlabs-mysql-5.1.0/spec/spec_helper_local.rb    2018-01-22 01:18:17.714262096 +0000
@@ -1,2 +1,11 @@
+require 'pry'
 require 'rspec-puppet-facts'
 include RspecPuppetFacts
+
+add_custom_fact :root_home, '/root'
+add_custom_fact :mysql_solaris,
+  {
+    'major_dot_minor' => '5.7',
+    'major_minor' => '57',
+    'basedir' => '/usr/mysql/5.7'
+  }
diff --git a/spec/unit/facter/mysql_server_id_spec.rb b/spec/unit/facter/mysql_server_id_spec.rb
index a2e9fdf..e690246 100644
--- a/spec/unit/facter/mysql_server_id_spec.rb
+++ b/spec/unit/facter/mysql_server_id_spec.rb
@@ -11,7 +11,7 @@ describe Facter::Util::Fact do
         Facter.fact(:macaddress).stubs(:value).returns('3c:97:0e:69:fb:e1')
       end
       it do
-        Facter.fact(:mysql_server_id).value.to_s.should == '4116385'
+        expect(Facter.fact(:mysql_server_id).value.to_s).to eq('4116385')
       end
     end
 
@@ -20,7 +20,7 @@ describe Facter::Util::Fact do
         Facter.fact(:macaddress).stubs(:value).returns('00:00:00:00:00:00')
       end
       it do
-        Facter.fact(:mysql_server_id).value.to_s.should == '0'
+        expect(Facter.fact(:mysql_server_id).value.to_s).to eq('0')
       end
     end
   end
diff --git a/spec/unit/facter/mysql_solaris_spec.rb b/spec/unit/facter/mysql_solaris_spec.rb
new file mode 100644
index 0000000..811d810
--- /dev/null
+++ b/spec/unit/facter/mysql_solaris_spec.rb
@@ -0,0 +1,65 @@
+require 'spec_helper'
+
+describe Facter::Util::Fact do
+  before do
+    Facter.clear
+    Facter.stubs(:value).with(:kernel).returns('SunOS')
+    Facter.stubs(:value).with('os').returns(
+      {"name"=>"Solaris", "family"=>"Solaris", "release"=>{"major"=>"12",
+      "minor"=>"0", "full"=>"12.0"}}
+     )
+    # Stubbed osfamily fails without stubbed 'os'
+    Facter.stubs(:value).with(:osfamily).returns('Solaris')
+
+    # Stub facts we don't use that are now invoked somewhere
+    [:ec2_metadata, :hardwareisa, :processors ].each do |fact|
+      Facter.stubs(:value).with(fact)
+    end
+  end
+  context 'solaris' do
+    before do
+      Facter::Util::Resolution.stubs(:exec).
+        with('pkg mediator -H mysql').
+        returns('mysql        local     5.7     system')
+    end
+    context 'with mediator' do
+      context 'major_dot_minor' do
+        it {
+          expect(Facter.fact(:mysql_solaris).value['major_dot_minor']).to eq('5.7')
+        }
+      end
+      context 'basedir' do
+        it {
+          expect(Facter.fact(:mysql_solaris).value['basedir']).to eq('/usr/mysql/5.7')
+        }
+      end
+      context 'major_minor' do
+        it {
+          expect(Facter.fact(:mysql_solaris).value['major_minor']).to eq('57')
+        }
+      end
+    end
+    context 'without mediator' do
+      before do
+        Facter::Util::Resolution.stubs(:exec).
+          with('pkg mediator -H mysql').
+          returns(nil)
+      end
+      context 'major_dot_minor' do
+        it {
+          expect(Facter.fact(:mysql_solaris).value['major_dot_minor']).to eq('5.5')
+        }
+      end
+      context 'basedir' do
+        it {
+          expect(Facter.fact(:mysql_solaris).value['basedir']).to eq('/usr/mysql/5.5')
+        }
+      end
+      context 'major_minor' do
+        it {
+          expect(Facter.fact(:mysql_solaris).value['major_minor']).to eq('55')
+        }
+      end
+    end
+  end
+end
--- puppetlabs-mysql-5.1.0/spec/unit/facter/mysqld_version_spec.rb.orig 2018-01-22 01:19:02.735751646 +0000
+++ puppetlabs-mysql-5.1.0/spec/unit/facter/mysqld_version_spec.rb      2018-01-22 01:22:32.323897905 +0000
@@ -1,18 +1,13 @@
 require 'spec_helper'

 describe Facter::Util::Fact do
-  before(:each) do
-    Facter.clear
-  end
+  before { Facter.clear }

-  describe 'mysqld_version' do
-    context 'with value' do
-      before :each do
-        Facter::Util::Resolution.stubs(:exec).with('mysqld -V 2>/dev/null').returns('mysqld  Ver 5.5.49-37.9 for Linux on x86_64 (Percona Server (GPL), Release 37.9, Revision efa0073)')
-      end
-      it {
-        expect(Facter.fact(:mysqld_version).value).to eq('mysqld  Ver 5.5.49-37.9 for Linux on x86_64 (Percona Server (GPL), Release 37.9, Revision efa0073)')
-      }
-    end
+  context 'mysqld_version' do
+    before do
+    it {
+      Facter::Util::Resolution.stubs(:exec).with('mysqld -V 2>/dev/null').returns('mysqld  Ver 5.5.49-37.9 for Linux on x86_64 (Percona Server (GPL), Release 37.9, Revision efa0073)')
+      expect(Facter.fact(:mysqld_version).value).to eq('mysqld  Ver 5.5.49-37.9 for Linux on x86_64 (Percona Server (GPL), Release 37.9, Revision efa0073)')
+    }
   end
 end
-- 
2.14.1

